File: src/services/auth.service.ts
import { userRepository } from "@/repositories";
import type { Role, UserEntity, UserRepository } from "@/repositories/contracts";
import { query } from "@/db/raw";
import { hashPassword, verifyPassword } from "@/lib/password";
import { signToken } from "@/lib/jwt";
import { ApiError } from "@/lib/http";

export interface AuthResult {
  user: Omit<UserEntity, "passwordHash">;
  token: string;
}

export class AuthService {
  constructor(private readonly users: UserRepository = userRepository) {}

  async register(input: {
    name: string;
    email: string;
    password: string;
    phone?: string;
    role?: string;
  }): Promise<AuthResult> {
    const email = input.email.trim().toLowerCase();
    if (await this.users.existsByEmail(email)) {
      throw new ApiError(409, "An account with this email already exists", "EMAIL_TAKEN");
    }

    const role = this.normalizeRole(input.role ?? "customer");

    const user = await this.users.create({
      name: input.name.trim(),
      email,
      passwordHash: await hashPassword(input.password),
      phone: input.phone ?? null,
      role,
    });

    // A rider registration also creates a rider profile record.
    if (role === "rider") {
      await query(`INSERT INTO delivery_riders (user_id) VALUES ($1)`, [user.id]);
    }

    return this.toAuthResult(user);
  }

  async login(email: string, password: string): Promise<AuthResult> {
    const user = await this.users.findByEmail(email.trim().toLowerCase());
    if (!user || !user.passwordHash || !(await verifyPassword(password, user.passwordHash))) {
      throw new ApiError(401, "Invalid email or password", "INVALID_CREDENTIALS");
    }
    if (user.isSuspended) {
      throw new ApiError(403, "Your account has been suspended. Please contact support.", "ACCOUNT_SUSPENDED");
    }
    return this.toAuthResult(user);
  }

  private normalizeRole(role: string): Role {
    const valid: Role[] = ["customer", "restaurant", "rider", "learner", "mentor"];
    if (valid.includes(role as Role)) return role as Role;
    if (role === "admin") throw new ApiError(403, "Admin accounts cannot be self-registered", "FORBIDDEN");
    throw new ApiError(400, "Invalid role", "INVALID_ROLE");
  }

  private async toAuthResult(user: UserEntity): Promise<AuthResult> {
    const token = await signToken(String(user.id), { role: user.role, email: user.email });
    const { passwordHash: _hash, ...safe } = user;
    return { user: safe, token };
  }
}

export const authService = new AuthService();

File: backend/src/main/java/com/foodrush/service/AuthService.java
package com.foodrush.service;

import com.foodrush.common.ApiException;
import com.foodrush.common.Role;
import com.foodrush.dto.Dtos;
import com.foodrush.entity.DeliveryRider;
import com.foodrush.entity.User;
import com.foodrush.repository.DeliveryRiderRepository;
import com.foodrush.repository.UserRepository;
import com.foodrush.security.JwtService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Map;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final DeliveryRiderRepository riderRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    @Transactional
    public Map<String, Object> register(Dtos.RegisterRequest req) {
        String email = req.email().trim().toLowerCase();
        if (userRepository.existsByEmail(email)) {
            throw ApiException.conflict("An account with this email already exists", "EMAIL_TAKEN");
        }

        Role role = Role.CUSTOMER;
        if (req.role() != null && !req.role().isBlank()) {
            try {
                role = Role.valueOf(req.role().trim().toUpperCase());
            } catch (IllegalArgumentException e) {
                throw ApiException.badRequest("Invalid role", "INVALID_ROLE");
            }
            if (role == Role.ADMIN) {
                throw ApiException.forbidden("Admin accounts cannot be self-registered", "FORBIDDEN");
            }
        }

        User user = userRepository.save(User.builder()
                .name(req.name().trim())
                .email(email)
                .passwordHash(passwordEncoder.encode(req.password()))
                .phone(req.phone())
                .role(role)
                .build());

        if (role == Role.RIDER) {
            riderRepository.save(DeliveryRider.builder().userId(user.getId()).build());
        }

        return Map.of("user", Dtos.UserResponse.from(user), "token", jwtService.generateToken(user));
    }

    public Map<String, Object> login(Dtos.LoginRequest req) {
        User user = userRepository.findByEmail(req.email().trim().toLowerCase())
                .orElseThrow(() -> ApiException.unauthorized("Invalid email or password", "INVALID_CREDENTIALS"));

        if (!passwordEncoder.matches(req.password(), user.getPasswordHash())) {
            throw ApiException.unauthorized("Invalid email or password", "INVALID_CREDENTIALS");
        }
        if (user.isSuspended()) {
            throw ApiException.forbidden("Your account has been suspended. Please contact support.", "ACCOUNT_SUSPENDED");
        }

        return Map.of("user", Dtos.UserResponse.from(user), "token", jwtService.generateToken(user));
    }
}
